pkg/networkserver/redis/scheduled_downlink_matcher.go

```diff
@@ -0,0 +1,87 @@
+// Copyright © 2021 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package redis
+
+import (
+	"context"
+	"regexp"
+
+	"github.com/go-redis/redis/v8"
+	"go.thethings.network/lorawan-stack/v3/pkg/errors"
+	"go.thethings.network/lorawan-stack/v3/pkg/networkserver/internal/time"
+	ttnredis "go.thethings.network/lorawan-stack/v3/pkg/redis"
+	"go.thethings.network/lorawan-stack/v3/pkg/ttnpb"
+)
+
+// storedScheduledDownlinkTTL is the duration before metadata for scheduled downlinks expire.
+// TxAcknowledgements received after this TTL will be considered invalid and will not match any downlinks.
+const storedScheduledDownlinkTTL = time.Minute
+
+// ScheduledDownlinkMatcher is an implementation of networkserver.ScheduledDownlinkMatcher.
+type ScheduledDownlinkMatcher struct {
+	Redis *ttnredis.Client
+}
+
+func (m *ScheduledDownlinkMatcher) cidToKey(cid string) string {
+	return m.Redis.Key("cid", cid)
+}
+
+var parseDownlinkIDRegex = regexp.MustCompile(`^ns:downlink:([0-9a-zA-Z]+)$`)
```

Comment:
This is the edgiest case, but what if someone intentionally adds `ns:downlink:mystaticthinghere` as a correlation ID ? Since the order of the correlation IDs is not defined, this may act weirder than it seems.

I think hashing the raw payload would be better, but I'm open to suggestions.